### ISSUE TITLE
Add  System.Collections.Concurrent.ConcurrentHashSet<T>

### ISSUE THREAD

**Issue by LeaFrock** [dotnet/runtime#39919]
## Background and Motivation

We already have ConcurrentDictionary which takes great advantages at concurrent environment comparing to Dictionary. However, HashSet doesn't have a **Concurrent** version.
Sometimes(maybe very often in fact), we hope to use **ConcurrentHashSet** easily which its elements are never duplicated, with an O(1) operation for contains or remove methods at concurrent environment. It's quite different from ConcurrentBag and provides another choice for suitable usage.
You can also see details at issue #16443 . Almost 5 years passes, nobody gives an API proposal, so I'm here to finish it.

## Proposed API

``` C#

namespace System.Collections.Concurrent
{
    public class ConcurrentHashSet<T> : ICollection<T>, ISet<T>, IReadOnlySet<T>
    {
         public int Count { get; }
         public System.Collections.Generic.IEqualityComparer<T> Comparer { get; }

         public bool TryAdd(T item);
         public bool TryGetValue(T equalValue, out T actualValue);
         public T GetOrAdd(T item);
         public void Clear();
         public bool Contains(T item);
         public bool Remove(T item);

         public void CopyTo(T[] array);
         public void CopyTo(T[] array, int arrayIndex);
         public void CopyTo(T[] array, int arrayIndex, int count);

         public Enumerator GetEnumerator ();
    }
}
```

## Usage Examples

The **main goal** of ConcurrentHashSet is to improve performance and reduce cost, taking replace of HashSet-With-Lock pattern for concurrent environment. If we have it, the code can be written like below,

``` C#

public class Sample()
{
     // private readonly object _syncRoot = new object();

     private readonly ConcurrentHashSet<Guid> _concurrentGuidSet = new ConcurrentHashSet<Guid>();
     // private readonly HashSet<Guid> _guidSet = new Hash<Guid>();

     public async Task Generate()
     {
          for(int i = 0; i < 100; i++)
          {
               Task.Run( () =>
               {
                    // The goal is to improve the performance, and ease the way of programming here.
                    _concurrentGuidSet.TryAdd(Guid.NewGuid());
                    //lock(_syncRoot)
                    //{
                    //     _guidSet.Add(Guid.NewGuid());
                    //}
               })
          }
     }
}

```

The difference could be very similiar to the one between Dictionary and ConcurrentDictionary.

## Alternative Designs

1. Many developers tell that they use `ConcurrentDictionary<T, byte>` as a replace when meeting the requirement. They don't want the `Value` actually which is meaningless and totally a waste of memory. However, someones suggest that the first version of ConcurrentHashSet can be implemented by underlying `ConcurrentDictionary<T, byte>`, in order to add the APIs as soon as possible. Individually, I prefer to an independent implementation which is necessary to be discussed between excellent engineers of Microsoft & our community.

2. Though it's the concurrent version of HashSet, I'm not sure whether each API of HashSet needs a copy, such as ExceptWith/IsSubsetOf/Overlaps/RemoveWhere/UnionWith...In my opinion, not all APIs make sense while running in a concurrent environment. Of course, the discussion is required.

## Risks

It might not own concurrent version of the whole APIs of HashSet. Because it's a new collection, no break changes will occur. It just needs a pretty design :).

**Comment by LeaFrock** [comment]
By the way, this issue is not duplicated of #15326 .

**Comment by danmosemsft** [comment]
Thanks for doing this. Could you please update the proposal to be complete? No implementation is needed, but everything that would be public, eg.,

- all constructors
- all members fulfiling the interface contracts - whether explicit or implicit. Eg., there's no getter for the generic enumerator, etc.
- to your question in #2, does it need ISet<T> and IReadOnlySet<T>-- it sounds like your use case is for storage and retrieval, not set operations like intersect and superset. Better to mimimally satisfy demonstrated needs, the interfaces can always be added later without breaking code, if use case emerges. 
- CopyTo probably should be hidden in explicit interface implementation, as for ConcurrentDictionary
- Remove should probably be TryRemove
- do we need XXAdd overloads with a factory delegate, as for ConcurrentDictionary?
- do we need AddOrUpdate?
- do we need TryUpdate?

**Comment by LeaFrock** [comment]
@danmosemsft  Updated, thanks!

> * all constructors

Just take reference of `ConcurrentDictionary<TKey, TValue>`.

> * all members fulfiling the interface contracts - whether explicit or implicit. Eg., there's no getter for the generic enumerator, etc.
> * CopyTo probably should be hidden in explicit interface implementation, as for ConcurrentDictionary

Sorry, I'm confused whether the whole members of interface contract should be listed above? I deleted overloads of `CopyTo` and added getter for the generic enumerator.

> * to your question in #2, does it need ISet and IReadOnlySet-- it sounds like your use case is for storage and retrieval, not set operations like intersect and superset. Better to mimimally satisfy demonstrated needs, the interfaces can always be added later without breaking code, if use case emerges.

Agreed. I moved the related content to 'Alternative Designs'.

> * Remove should probably be TryRemove

Agreed & done.

> * do we need XXAdd overloads with a factory delegate, as for ConcurrentDictionary?
> * do we need AddOrUpdate?
> * do we need TryUpdate?

I think **NO**, because it's a simpler collection than ConcurrentDictionary. For ConcurrentDictionary, we usually use AddOrUpdate/TryUpdate or factory delegate overloads to update `Value` only. ConcurrentHashSet owns the `Key` part only, so I don't see any point in them. Any comments welcomed.

> * does it need IsEmpty? this is syntactic sugar that with present JIT may be less efficient than checking Count == 0. 

I think we can add it. The following is the reasons from the perspective of my view, 
* ConcurrentDictionary owns it
* it really simplify codes
* the performance gap is acceptable and I believe JIT performance here will be optimized sooner or later

**Comment by danmosemsft** [comment]
Thanks 

> Sorry, I'm confused whether the whole members of interface contract should be listed above? 

We're looking for the same format as in this file 
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Collections.Concurrent/ref/System.Collections.Concurrent.cs#L75

**Comment by LeaFrock** [comment]
@danmosemsft Got it. I'll finish it at another time.

**Comment by danmosemsft** [comment]
Ok thanks. The owners of this area will take a look. It might be a little while since we are focused on shipping 5.0 right now.

**Comment by eiriktsarpalis** [comment]
> Many developers tell that they use ConcurrentDictionary<T, byte> as a replace when meeting the requirement. They don't want the Value actually which is meaningless and totally a waste of memory.

This is a compromise that has more or less worked fine, at least for my own requirements in the past. It would be interesting if we could quantify any performance improvements that come out of a dedicated ConcurrentHashSet implementation.

**Comment by bennidhamma** [comment]
Even if you just used a `ConcurrentDictionary<T, byte>` initially, under the covers, people could start using a cleaner interface.

**Comment by eiriktsarpalis** [comment]
> Even if you just used a ConcurrentDictionary<T, byte> initially, under the covers, people could start using a cleaner interface.

Adding new classes always comes at a cost, even if they're just facades.

**Comment by LeaFrock** [comment]
> It would be interesting if we could quantify any performance improvements that come out of a dedicated ConcurrentHashSet implementation.

Well, so the situation becomes that we're all waiting for some kind of implementation by someone now, in order to compare it with the current solution... that doesn't sound good =.=

**Comment by danmosemsft** [comment]
It will be more space efficient, even if it's no faster, which seems potentially worthwhile to me.

I noticed this (didn't look at it): https://github.com/i3arnon/ConcurrentHashSet by @i3arnon

**Comment by eiriktsarpalis** [comment]
In practice though we'd be looking at maybe shaving one byte per [`Node` instance](https://github.com/dotnet/runtime/blob/e29b83934e7ceef9905f82bd6328cbc72db2278b/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L2165) which I'm not sure would even register in heap allocations. That being said, there might be opportunities for other optimizations as well, happy to be surprised :-)

**Comment by danmosemsft** [comment]
That is true.

**Comment by stephentoub** [comment]
> In practice though we'd be looking at maybe shaving one byte per Node instance

On 64-bit it won't even save that.

**Comment by i3arnon** [comment]
> On 64-bit it won't even save that.

It depends on the type being used, doesn't it?
I did some naive tests with benchmark-net compared with `ConcurrentHashSet` and I can see some difference when `Tkey` is `int`, but not with reference types (on 64-bit) or any other sized types (`long`, `short`, etc.)
I assume that's because of how `Node`'s members are being aligned. It consists of a reference to the next node (8 bytes), 2 `int`s (the key & hash-code, 8 bytes together) and the single byte value which requires another 8 byte block to keep alignment.

**Comment by stephentoub** [comment]
> It depends on the type being used, doesn't it?

Yes. The node stores the key,  the value,  a next node reference,  and an int hash code, so whether the value increases the size of the object depends on whether the key and value need any padding. If the value is byte, then if the key is any reference type, for example,  the byte value will just be in what is otherwise padding space from the hashcode.
